WordPress/Classes/Services/MediaUploadCoordinator.swift

```diff
@@ -0,0 +1,147 @@
+import Foundation
+
+/// MediaUploadCoordinator is responsible for creating and uploading new media
+/// items, independently of a specific view controller. It should be accessed
+/// via the `shared` singleton.
+///
+class MediaUploadCoordinator {
+    static let shared = MediaUploadCoordinator()
+
+    private let queue = DispatchQueue(label: "org.wordpress.mediauploadcoordinator")
+
+    // MARK: - Adding Media
+
+    /// Adds the specified media asset to the specified blog. The upload process
+    /// can be observed by adding an observer block using the `addObserver(_:for:)` method.
+    ///
+    /// - parameter asset: The asset to add.
+    /// - parameter blog: The blog that the asset should be added to.
+    ///
+    func addMedia(from asset: ExportableAsset, to blog: Blog) {
+        guard let asset = asset as? PHAsset else {
+            return
+        }
+
+        let context = ContextManager.sharedInstance().mainContext
+        let service = MediaService(managedObjectContext: context)
+        service.createMedia(with: asset,
+                            forBlogObjectID: blog.objectID,
+                            thumbnailCallback: nil,
+                            completion: { media, error in
+                                guard let media = media else {
+                                    return
+                                }
+
+                                self.begin(media)
+
+                                var progress: Progress? = nil
+                                service.uploadMedia(media,
+                                                    progress: &progress,
+                                                    success: {
+                                                        self.end(media)
+                                }, failure: { _ in
+                                    self.end(media)
+                                })
+        })
+    }
+
+    // MARK: - Observing
+
+    typealias ObserverBlock = (Media, MediaState) -> Void
+
+    private var mediaObservers = [UUID: MediaObserver]()
+
+    /// Add an observer to receive updates when media items are updated.
+    ///
+    /// - parameter onUpdate: A block that will be called whenever media items
+    ///                       (or a specific media item) are updated.
+    /// - parameter media: An optional specific media item to receive updates for.
+    ///                    If provided, the `onUpdate` block will only be called
+    ///                    for updates to this media item, otherwise it will be
+    ///                    called when changes occur to _any_ media item.
+    /// - returns: A UUID that can be used to unregister the observer block at a later time.
+    ///
+    func addObserver(_ onUpdate: @escaping ObserverBlock, for media: Media? = nil) -> UUID {
```

Comment:
We probably should refer on the docs here, or in the class docs in what thread the observer blocks will be invoked.

Reply:
Thanks!